ports/stm32f4/common-hal/busio/SPI.c

```diff
@@ -23,37 +24,320 @@
+    self->handle.Init.NSS = SPI_NSS_SOFT;
+    self->handle.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_16;
+    self->handle.Init.FirstBit = SPI_FIRSTBIT_MSB;
+    self->handle.Init.TIMode = SPI_TIMODE_DISABLE;
+    self->handle.Init.CRCCalculation = SPI_CRCCALCULATION_DISABLE;
+    self->handle.Init.CRCPolynomial = 10;
+    if (HAL_SPI_Init(&self->handle) != HAL_OK)
+    {
+        mp_raise_RuntimeError(translate("SPI Init Error"));
+    } else {
+        mp_printf(&mp_plat_print, "Constructed! spi clock speed:%u\n",(HAL_RCC_GetPCLK2Freq()/16));
+    }
+
+    self->baudrate = (HAL_RCC_GetPCLK2Freq()/16);
+    self->prescaler = 16;
+    self->polarity = 0;
+    self->phase = 1;
+    self->bits = 8;
+
+    claim_pin(sck);
+    claim_pin(mosi);
+    claim_pin(miso);
 }
 
 void common_hal_busio_spi_never_reset(busio_spi_obj_t *self) {
 }
 
 bool common_hal_busio_spi_deinited(busio_spi_obj_t *self) {
-    return 0;
+    return self->sck->pin == mp_const_none;
 }
 
 void common_hal_busio_spi_deinit(busio_spi_obj_t *self) {
+    #ifdef SPI1
+    if(self->handle.Instance==SPI1) {
+        reserved_spi[0] = false;
+        __HAL_RCC_SPI1_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI2
+    if(self->handle.Instance==SPI2) {
+        reserved_spi[1] = false;
+        __HAL_RCC_SPI2_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI3
+    if(self->handle.Instance==SPI3) {
+        reserved_spi[2] = false;
+        __HAL_RCC_SPI3_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI4
+    if(self->handle.Instance==SPI4) {
+        reserved_spi[3] = false;
+        __HAL_RCC_SPI4_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI5
+    if(self->handle.Instance==SPI5) {
+        reserved_spi[4] = false;
+        __HAL_RCC_SPI5_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI6
+    if(self->handle.Instance==SPI6) {
+        reserved_spi[5] = false;
+        __HAL_RCC_SPI6_CLK_DISABLE(); 
+    }
+    #endif
+    reset_pin_number(self->sck->pin->port,self->sck->pin->number);
+    reset_pin_number(self->mosi->pin->port,self->mosi->pin->number);
+    reset_pin_number(self->miso->pin->port,self->miso->pin->number);
+    self->sck = mp_const_none;
+    self->mosi = mp_const_none;
+    self->miso = mp_const_none;
+}
+
+static uint32_t stm32_baud_to_spi_div(uint32_t baudrate, uint16_t * prescaler) {
+    static const uint32_t baud_map[8][2] = {
+        {2,SPI_BAUDRATEPRESCALER_2},
+        {4,SPI_BAUDRATEPRESCALER_4},
+        {8,SPI_BAUDRATEPRESCALER_8},
+        {16,SPI_BAUDRATEPRESCALER_16},
+        {32,SPI_BAUDRATEPRESCALER_32},
+        {64,SPI_BAUDRATEPRESCALER_64},
+        {128,SPI_BAUDRATEPRESCALER_128},
+        {256,SPI_BAUDRATEPRESCALER_256}
+    };
+    size_t i = 0;
+    uint16_t divisor;
+    do {
+        divisor = baud_map[i][0];
+        if (baudrate >= (HAL_RCC_GetPCLK2Freq()/divisor)) {
+            *prescaler = divisor;
+            return baud_map[i][1];
+        }
+        i++;
+    } while (divisor != 256);
+    //should never get here
```

Comment:
I think it can get here if the given baudrate is really slow (less than plck / 256) and that's ok. Just set it to 256.

Reply:
Do you want me to throw a python message in there or anything?

Reply:
Nope! Looks good as is.